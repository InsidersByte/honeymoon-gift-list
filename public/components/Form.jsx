/* @flow */

import React from 'react';
import Loader from './common/Loader';

type PropsType = {
    loading: boolean,
    saving: boolean,
    // FIXME should just be children not children?
    children?: React$Element<any> | Array<React$Element<any>>,
    onSubmit: Function,
};

export default function Form(props: PropsType) {
    const { loading, saving, children, onSubmit } = props;

    if (loading) {
        return (
            <Loader loading />
        );
    }

    return (
        <form {...props} onSubmit={onSubmit}>
            <fieldset disabled={saving} style={{ margin: 0, padding: 0, border: 'none' }}>
                {children}
            </fieldset>
        </form>
    );
}
