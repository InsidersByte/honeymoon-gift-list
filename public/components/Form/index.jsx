/* @flow */

import React from 'react';
import Loader from '../Loader';

type PropsType = {
  loading: boolean,
  saving: boolean,
  children?: React$Element<any> | Array<React$Element<any>>,
  onSubmit: (event: SyntheticEvent) => void,
};

const Form = ({ loading, saving, onSubmit, children, ...props }: PropsType) => {
  if (loading) {
    return <Loader loading />;
  }

  return (
    <form {...props} onSubmit={onSubmit}>
      <fieldset disabled={saving} style={{ margin: 0, padding: 0, border: 'none' }}>
        {children}
      </fieldset>
    </form>
  );
};

export default Form;
