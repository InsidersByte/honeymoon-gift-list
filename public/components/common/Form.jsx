import React from 'react';
import Loader from './Loader';

export default function Form(props) {
    return (
        <Loader loading={props.loading}>
            <form onSubmit={props.onSubmit}>
                <fieldset disabled={props.saving}>
                    {props.children}
                </fieldset>
            </form>
        </Loader>
    );
}

Form.propTypes = {
    onSubmit: React.PropTypes.func.isRequired,
    saving: React.PropTypes.bool.isRequired,
    loading: React.PropTypes.bool.isRequired,
    children: React.PropTypes.oneOfType([
        React.PropTypes.element,
        React.PropTypes.arrayOf(React.PropTypes.element),
    ]).isRequired,
};
