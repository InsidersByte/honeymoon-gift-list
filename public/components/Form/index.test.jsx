import React from 'react';
import { shallow } from 'enzyme';
import Form from './';

describe('Form', () => {
    describe('loading', () => {
        it('should render correctly', () => {
            const props = {
                loading: true,
            };

            const wrapper = shallow(<Form {...props} />);

            expect(wrapper).toMatchSnapshot();
        });

        it('should render Loader', () => {
            const props = {
                loading: true,
            };

            const wrapper = shallow(<Form {...props} />);
            const loader = wrapper.find('Loader');

            expect(loader.length).toBe(1);
            expect(loader.props()).toEqual({ loading: true });
        });
    });

    describe('not loading', () => {
        it('should render correctly', () => {
            const props = {
                loading: false,
                saving: false,
                onSubmit: jest.fn(),
            };

            const wrapper = shallow(<Form {...props}><h1>hello world</h1></Form>);

            expect(wrapper).toMatchSnapshot();
        });

        it('should render a form', () => {
            const props = {
                loading: false,
                saving: false,
                onSubmit: jest.fn(),
            };

            const wrapper = shallow(<Form {...props}><h1>hello world</h1></Form>);

            const form = wrapper.find('form');

            expect(form.length).toBe(1);
            expect(form.prop('onSubmit')).toEqual(props.onSubmit);
        });

        it('should render a fieldset', () => {
            const props = {
                loading: false,
                saving: false,
                onSubmit: jest.fn(),
            };

            const wrapper = shallow(<Form {...props}><h1>hello world</h1></Form>);

            const fieldset = wrapper.find('fieldset');

            expect(fieldset.length).toBe(1);
            expect(fieldset.prop('disabled')).toBe(false);
            expect(fieldset.prop('style')).toEqual({ margin: 0, padding: 0, border: 'none' });
        });

        it('should render children', () => {
            const props = {
                loading: false,
                saving: false,
                onSubmit: jest.fn(),
            };

            const wrapper = shallow(<Form {...props}><h1>hello world</h1></Form>);

            const fieldset = wrapper.find('fieldset');
            const children = fieldset.children();

            expect(children.html()).toBe('<h1>hello world</h1>');
        });

        it('should set the fieldset to disabled when saving', () => {
            const props = {
                loading: false,
                saving: true,
                onSubmit: jest.fn(),
            };

            const wrapper = shallow(<Form {...props}><h1>hello world</h1></Form>);

            const fieldset = wrapper.find('fieldset');

            expect(fieldset.length).toBe(1);
            expect(fieldset.prop('disabled')).toBe(true);
            expect(fieldset.prop('style')).toEqual({ margin: 0, padding: 0, border: 'none' });
        });

        it('should call onSubmit when submitting the form', () => {
            const props = {
                loading: false,
                saving: true,
                onSubmit: jest.fn(),
            };

            const wrapper = shallow(<Form {...props}><h1>hello world</h1></Form>);

            const form = wrapper.find('form');

            expect(props.onSubmit).not.toHaveBeenCalled();
            form.simulate('submit');
            expect(props.onSubmit).toHaveBeenCalledTimes(1);
        });
    });
});
