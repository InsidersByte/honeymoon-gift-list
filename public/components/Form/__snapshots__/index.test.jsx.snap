// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form loading should render correctly 1`] = `
<Loader
  loading={true}
/>
`;

exports[`Form not loading should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <fieldset
    disabled={false}
    style={
      Object {
        "border": "none",
        "margin": 0,
        "padding": 0,
      }
    }
  >
    <h1>
      hello world
    </h1>
  </fieldset>
</form>
`;
